File: js/src/matrix.js
const matrixSize = 10;

export function sumOfMatrixElements(matrix) {
  let sum = 0;
  for (let i = 0; i < matrixSize; i++) {
    for (let j = 0; j < matrixSize; j++) {
      sum += matrix[i][j];
    }
  }
  return sum;
}

/**
 * 
 * @param {*} matrix1 
 * @param {*} matrix2 
 * @param {*} index - We want to be sure that there is no internal optimization during testing 
 * @returns 
 */
export function multiplyMatrices(matrix1, matrix2, index) {
  const result = getEmptyMatrix()
  for (let i = 0; i < matrixSize; i++) {
    for (let j = 0; j < matrixSize; j++) {
      for (let k = 0; k < matrixSize; k++) {
        result[i][j] += matrix1[i][k] * matrix2[k][j] + index; // We add index to avoid internal optimization
      }
    }
  }
  return result
}

export function getEmptyMatrix() {
  return new Array(matrixSize).fill(0).map(() => new Array(matrixSize).fill(0));
}

export function fillSequencialyMatrix(matrix) {
  let value = 0;
  for (let i = 0; i < matrixSize; i++) {
    for (let j = 0; j < matrixSize; j++) {
      matrix[i][j] = value++;
    }
  }
  return matrix
}

File: rust/src/matrix.rs
pub fn sum_of_matrix_elements(matrix: &Vec<Vec<f64>>, matrix_size: usize) -> f64 {
    let mut sum = 0.0;
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            sum += matrix[i][j];
        }
    }
    sum
}


pub fn multiply_matrices(
    matrix1: &Vec<Vec<f64>>, 
    matrix2: &Vec<Vec<f64>>, 
    index: usize, 
    matrix_size: usize
) -> Vec<Vec<f64>> {
    let mut result = get_empty_matrix(matrix_size);
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            for k in 0..matrix_size {
                result[i][j] += matrix1[i][k] * matrix2[k][j] + index as f64;
            }
        }
    }
    result
}


pub fn get_empty_matrix(matrix_size: usize) -> Vec<Vec<f64>> {
    vec![vec![0.0; matrix_size]; matrix_size]
}

pub fn fill_sequential_matrix(matrix: &mut Vec<Vec<f64>>, matrix_size: usize) {
    let mut value = 0.0;
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            matrix[i][j] = value;
            value += 1.0;
        }
    }
}

pub fn sum_of_matrix_elements_i64(matrix: &Vec<Vec<i64>>, matrix_size: usize) -> i64 {
    let mut sum = 0;
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            sum += matrix[i][j];
        }
    }
    sum
}


pub fn multiply_matrices_i64(
    matrix1: &Vec<Vec<i64>>, 
    matrix2: &Vec<Vec<i64>>, 
    index: usize, 
    matrix_size: usize
) -> Vec<Vec<i64>> {
    let mut result = get_empty_matrix_i64(matrix_size);
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            for k in 0..matrix_size {
                result[i][j] += matrix1[i][k] * matrix2[k][j] + index as i64;
            }
        }
    }
    result
}


pub fn get_empty_matrix_i64(matrix_size: usize) -> Vec<Vec<i64>> {
    vec![vec![0; matrix_size]; matrix_size]
}

pub fn fill_sequential_matrix_i64(matrix: &mut Vec<Vec<i64>>, matrix_size: usize) {
    let mut value = 0;
    for i in 0..matrix_size {
        for j in 0..matrix_size {
            matrix[i][j] = value;
            value += 1;
        }
    }
}
